server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheDecodeContext.java

```diff
@@ -75,6 +90,182 @@ public RequestParameters getParams() {
+            counterManager.getStrongCounter(name).reset().whenComplete((aVoid, throwable) -> {
+               if (throwable != null) {
+                  checkCounterThrowable(sendResponse, throwable);
+               } else {
+                  sendResponse.accept(createEmptyResponse(header, OperationStatus.Success));
+               }
+            });
+            break;
+         case WEAK:
+            counterManager.getWeakCounter(name).reset().whenComplete((aVoid, throwable) -> {
+               if (throwable != null) {
+                  checkCounterThrowable(sendResponse, throwable);
+               } else {
+                  sendResponse.accept(createEmptyResponse(header, OperationStatus.Success));
+               }
+            });
+            break;
+      }
+   }
+
+   void counterAddAndGet(Consumer<Response> sendResponse) {
+      CounterAddDecodeContext decodeContext = operationContext();
+      String name = decodeContext.getCounterName();
+      CounterConfiguration config = counterManager.getConfiguration(name);
+      if (config == null) {
+         sendResponse.accept(missingCounterResponse());
+         return;
+      }
+      switch (config.type()) {
+         case UNBOUNDED_STRONG:
+         case BOUNDED_STRONG:
+            counterManager.getStrongCounter(name)
+                  .addAndGet(decodeContext.getValue())
+                  .whenComplete((value, throwable) -> {
+                     if (throwable != null) {
+                        checkCounterThrowable(sendResponse, throwable);
+                     } else {
+                        sendResponse.accept(new CounterValueResponse(header, value));
+                     }
+                  });
+            break;
+         case WEAK:
+            counterManager.getWeakCounter(name)
+                  .add(decodeContext.getValue())
+                  .whenComplete((aVoid, throwable) -> {
+                     if (throwable != null) {
+                        checkCounterThrowable(sendResponse, throwable);
+                     } else {
+                        sendResponse.accept(new CounterValueResponse(header, 0));
+                     }
+                  });
+            break;
+      }
+   }
+
+   Response getCounterConfiguration() {
+      String counterName = operationContext();
+      CounterConfiguration configuration = counterManager.getConfiguration(counterName);
+      return configuration == null ?
+             missingCounterResponse() :
+             new CounterConfigurationResponse(header, configuration);
+   }
+
+   Response isCounterDefined() {
+      String counterName = operationContext();
+      boolean result = counterManager.isDefined(counterName);
+      return createCounterBooleanResponse(result);
+   }
+
+   Response createCounter() {
+      CounterCreateDecodeContext decodeContext = operationContext();
+      boolean result = counterManager.defineCounter(decodeContext.getCounterName(), decodeContext.getConfiguration());
```

Comment:
`defineCounter` involves RPC and is synchronous, while the decoder should not block. You need to add `defineCounterAsync` to the CounterManager.

Reply:
I'm confused!
I don't think this method is invoked so often that it deserves an async approach. 
Also, all the cache operations are sync and use at least one RPC.

Reply:
Ouch, the cache ops are really still sync! 😭 I thought this was already rewritten. It's not a matter of 'how often'. But due to other syncs, probably not really needed at this stage but the time shall come...

Reply:
I created async method to the `EmbeddedCounterManager` but I kept the `CounterManager` API sync.  Only the `remove()` method is sync because I didn't find an easy way around the `ConcurrentHashMap.compute()`